tools/src/launcher/Main.hs

```diff
@@ -517,11 +527,70 @@ spawnNode (path, args, mbLogPath) = do
             logInfo "Node has started"
             return (ph, asc)
 
-runWallet :: (FilePath, [Text]) -> M ExitCode
-runWallet (path, args) = do
-    logNotice "Starting the wallet"
-    view _1 <$>
-        liftIO (readProcessWithExitCode path (map toString args) mempty)
+runWallet :: Bool -> (FilePath, [Text], Maybe FilePath) -> M ExitCode
+runWallet shouldLog (path, args, mLogPath) = do
+    customLogger LInfo NWallet "Starting the wallet"
+    phvar <- newEmptyMVar
+    liftIO $ case mLogPath of
+        Just lp -> do
+            createDirectoryIfMissing True (directory lp)
+            (_, logHandle) <- (lp,) <$> openFile lp AppendMode
+            liftIO $ IO.hSetBuffering logHandle IO.LineBuffering
+            let cr = createProcHandle path args logHandle
+            system' phvar cr mempty
+        Nothing ->
+           if shouldLog then
+               liftIO $ do
+               let cr = createProcInherit path args
+               system' phvar cr mempty
+           else
+              liftIO $ do
+              (_, _, _, pid) <- runInteractiveProcess path (map toString args) Nothing Nothing
+              waitForProcess pid
+
+createProcInherit :: FilePath -> [Text] -> Process.CreateProcess
+createProcInherit path args =
+    (Process.proc (toString path) (map toString args))
+        { Process.std_in = Process.CreatePipe
+        , Process.std_out = Process.Inherit
+        , Process.std_err = Process.Inherit
+        }
+
+createProcHandle :: FilePath -> [Text] -> Handle -> Process.CreateProcess
+createProcHandle path args pHandle =
+    (Process.proc (toString path) (map toString args))
+        { Process.std_in = Process.CreatePipe
+        , Process.std_out = Process.UseHandle pHandle
+        , Process.std_err = Process.UseHandle pHandle
+        }
+
+createProcPipe :: FilePath -> [Text] -> Process.CreateProcess
+createProcPipe path args =
+    (Process.proc (toString path) (map toString args))
+        { Process.std_in = Process.CreatePipe
+        , Process.std_out = Process.CreatePipe
+        , Process.std_err = Process.CreatePipe
+        }
+
+asyncLog :: Handle -> Handle -> ProcessHandle -> NodeType -> IO ExitCode
```

Comment:
Are we really sure we need this function? It reads output of the spawned process and sends it to our stdout/stderr.
The only difference between this approach and inheriting stdout/stderr is that we can modify those messages. However, some messages take several lines. For example, block header takes several lines. With this approach it will look like:
```
[node:ERROR:ThreadId 4] [2017-12-07 16:55:41.20 UTC] MainBlockHeader:
[node:ERROR:ThreadId 4] [2017-12-07 16:55:41.20 UTC]     hash: 1f19aa3d840b964c55b1ac5cc3e4e19c8d3fa5b066fecedcdb2d811bb0f8d2b3
[node:ERROR:ThreadId 4] [2017-12-07 16:55:42.02 UTC]     previous block: f9f06e9ca71b5fc5c26e7c893e6d6f0011f35a262aa1f2959f05b2e3d0a808a6
…
```

Also it involves extra overhead and extra code.
I personally vote for removing it.

Reply:
@georgeee  asked to have `[node]` and `[wallet]` prepended to log output. If you think there is better solution I will gladly implement it.

Reply:
Frankly, I'd like to avoid printing output of two executables to the same stream. I think it will be mess. Some messages take several lines. If outputs of two executables are combined, these messages may be mixed and it will be hard to read them.
`asyncLog` approach assumes that each message is 1 line. But it's not the case.
Moreover, after code review I see that it's easy to misuse `asyncLog`, for instance by applying to the process with inherited stdout (which leads to infinite loop).
And in fact there are no cases when we need to print output of both wallet and node to console. We only need node to be able to log to console.
It would be easier for me to review if this function didn't exist :)
But if you want to leave it, then ok, just use it properly. It should be used only when both wallet and node print to console. It means that `loWalletLogging` is `True`, `loWalletLogPath` is `Nothing` and `loNodeLogPath` is `Nothing`. Currently it's used unconditionally by `system'` which is wrong.